extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/StaticTenantResolver.java

```diff
@@ -0,0 +1,267 @@
+            addPath(DEFAULT_TENANT, defaultTenant.oidcConfig(), builder);
+            for (Map.Entry<String, TenantConfigContext> e : staticTenantsConfig.entrySet()) {
+                addPath(e.getKey(), e.getValue().oidcConfig(), builder);
+            }
+            return builder.hasPaths() ? new PathMatchingTenantResolver(builder.build()) : null;
+        }
+
+        @Override
+        public String resolve(RoutingContext context) {
+            String tenantId = staticTenantPaths.match(context.normalizedPath()).getValue();
+            if (tenantId != null) {
+                LOG.debugf(
+                        "Tenant id '%s' is selected on the '%s' request path", tenantId, context.normalizedPath());
+                return tenantId;
+            }
+            return null;
+        }
+
+        private static ImmutablePathMatcher.ImmutablePathMatcherBuilder<String> addPath(String tenant, OidcTenantConfig config,
+                ImmutablePathMatcher.ImmutablePathMatcherBuilder<String> builder) {
+            if (config != null && config.tenantPaths.isPresent()) {
+                for (String path : config.tenantPaths.get()) {
+                    builder.addPath(path, tenant);
+                }
+            }
+            return builder;
+        }
+    }
+
+    private static final class IssuerBasedTenantResolver {
+
+        private final TenantConfigContext[] tenantConfigContexts;
+        private final boolean detectedTenantWithoutMetadata;
+
+        private IssuerBasedTenantResolver(TenantConfigContext[] tenantConfigContexts, boolean detectedTenantWithoutMetadata) {
+            this.tenantConfigContexts = tenantConfigContexts;
+            this.detectedTenantWithoutMetadata = detectedTenantWithoutMetadata;
+        }
+
+        public Uni<String> resolveTenant(RoutingContext context) {
+            return resolveTenant(context, 0);
+        }
+
+        private Uni<String> resolveTenant(RoutingContext context, int index) {
+            if (index == tenantConfigContexts.length) {
+                return Uni.createFrom().nullItem();
+            }
+            var tenantContext = tenantConfigContexts[index];
+            if (detectedTenantWithoutMetadata) {
+                // this is static tenant that didn't have OIDC metadata available at startup
+
+                if (tenantContext.getOidcMetadata() == null) {
+                    if (tenantContext.ready()) {
+                        return resolveTenant(context, index + 1);
+                    }
+
+                    return tenantContext.initialize()
+                            .onItemOrFailure()
+                            .transformToUni(new BiFunction<TenantConfigContext, Throwable, Uni<? extends String>>() {
+                                @Override
+                                public Uni<String> apply(TenantConfigContext newContext, Throwable throwable) {
+                                    if (throwable != null) {
```

Comment:
@michalvavrik If the context fails to initialize again here then it can continue for every request.

The same can probably happen in other cases.

Should have have a tenant config context status enum, instead of boolean, `NOT_INITIALIZED`, `INITIALIZED` and `FAILED_TO_INITIALIZE`, if it is `FAILED_TO_INITIALIZE` then we never try to re-initalize it, i.e, we only try to move from a possible  `NOT_INITIALIZED` to `INITIALIZED` only once, when trying to recover a failed context at the runtime for the first time... 

What do you think ?

Reply:
Trying to recover a failed static context at runtime already makes it 2 attempts for the static context, we can't continue indefinitely, it can make a system nearly irresponsive if we have more than one unreachable static context

Reply:
>  If the context fails to initialize again here then it can continue for every request. The same can probably happen in other cases.

My thinking was that if the tenant is not supposed to get ready, why is it not disabled? And if it can get ready, we don't know when (after 5 minutes, 10 minutes). I agree , this is can be an issue.

> Should we have a tenant config context status enum, instead of boolean, NOT_INITIALIZED, INITIALIZED and FAILED_TO_INITIALIZE, if it is FAILED_TO_INITIALIZE then we never try to re-initalize it, i.e, we only try to move from a possible NOT_INITIALIZED to INITIALIZED only once, when trying to recover a failed context at the runtime for the first time...

That is alright if you believe that cannot recover later. Can this be that pod with Quarkus app spins (is scaled) exactly when IdM is not available? You could say that they should keep IdM available I guess.

What about I change `ready` flag to `enum` that you proposed but user can choose if they want to keep trying even if it failed to initialize? It will require additional property, something like `quarkus.oidc.resolve-tenants-with-issuer.retry-tenant-initialization`? Or we could provide a bean that users can define (strategy that has one method `tryToInitialize`? And default strategy would be don't try after failed initialization.

If you don't mind, I don't think that enum is required outside of the issuer-based tenant resolver. And for flag I don't need synchronization, because initialization only succeeds when is ready. I'd like to avoid enum in favor of some counter inside of the issuer-based resolver. But this is a small thing, I can do the enum if you can see other usages?

Reply:
>My thinking was that if the tenant is not supposed to get ready, why is it not disabled? 

As far as I know, in most cases, these are test situations, or some transient initial connectivity issues.

>What about I change ready flag to enum that you proposed but user can choose if they want to keep trying even if it failed to initialize?

IMHO it is too flexible, the requests have started coming in, the system supporting the security access is either available now or not, and we don't do it for dynamic tenants. May be we can do it later for all types of tenants, keep retrying optionally...

>If you don't mind, I don't think that enum is required outside of the issuer-based tenant resolver.

Would the same be the case with any other static tenant resolver, which just returns a tenant id ? The code then will try to initialize, if it fails, the error is returned, and then it will start over from the next request...

Reply:
I don't have strong opinion, but this could be difference between good feature and minor fix. Please let's try another round:

> As far as I know, in most cases, these are test situations, or some transient initial connectivity issues.

ok, so sure, the tests can be fixed with this PR without retrying

> IMHO it is too flexible, the requests have started coming in, the system supporting the security access is either available now or not, and we don't do it for dynamic tenants. May be we can do it later for all types of tenants, keep retrying optionally...

no flag, no problem. What about:

```
/**
* By default Quarkus Issuer-based tenant resolver retries just once to initialize static tenant that wasn't available on the application startup. If your OIDC server is frequently available, you can implement this interface as a CDI bean.
*/
public interface StaticTenantInitializationStrategy {
     boolean initializeTenant(String tenantId);
}
```
?

> Would the same be the case with any other static tenant resolver, which just returns a tenant id ? The code then will try to initialize, if it fails, the error is returned, and then it will start over from the next request...

I think that is different scenario. What happens now (correct me if I am wrong):

- tenant resolver selects tenants independently whether they are enabled or not; e.g. based on path or end of the path (default tenant)
- Quarkus gets selected tenant that is supposed to be initialized
- Quarkus always tries to reinitialize it, not just once or twice, always

But what issuer-based tenant resolver does is more dangerous:

- let's try to initialize this tenant in case it is the tenant with my issuer (AKA: I might not need this tenant)

So this `May be we can do it later for all types of tenants, keep retrying optionally...` doesn't fit, maybe you could elaborate how you mean it.

Reply:
Re `StaticTenantInitializationStrategy` and the last question, re `all types of tenants`, why would we do possibly unlimited retries for those whose configuration is declared in `application.properties` but not those where the configuration is created dynamically, see what I mean ?
Also, we have never had anyone asking for this feature. Recovery, yes, but otherwise users would be getting empty screens in prod (401), tests will be failing anyway, unless they have been rewritten to loop, etc.

>Quarkus always tries to reinitialize tenant that is not ready, not just once or twice, always

Right, to be honest, I've never thought, that it would also be trying to do it, it just became obvious now :-). 

>let's try to initialize this tenant in case it is the tenant with my issuer (AKA: I might not need this tenant)

Indeed, with the issuer based resolution, a single loop can end up initializing all not-ready tenants, but all these tenants are meant to be initialized anyway, and in a busy system, it may well be concurrent

Reply:
> Re StaticTenantInitializationStrategy and the last question, re all types of tenants, why would we do possibly unlimited retries for those whose configuration is declared in application.properties but not those where the configuration is created dynamically, see what I mean ?

Sorry, I don't get a point yet. The strategy is only useful for the issuer-based tenant resolver. Because we do unlimited retries for those whose configuration is declared in application.properties when some tenant resolver selects a tenant that is not ready.

> Also, we have never had anyone asking for this feature. 

Fair point.

> and in a busy system, it may well be concurrent

AFAICT not a problem


> Indeed, with the issuer based resolution, a single loop can end up initializing all not-ready tenants, but all these tenants are meant to be initialized anyway

I agree, but it seems to contradict what you try to initialize just once after failed startup initialization. Just so that it is clear, conclusion is that we should retry just once after initialization on startup failed, correct?

I can do that, just want to confirm before I start.

Reply:
>The strategy is only useful for the issuer-based tenant resolver

But it would require us introducing a strategy specific interface, IMHO it does not really matter if this initialization attempt was attempted from the issuer based resolver ( FYI, I only referred to the concurrency to alleviate a concern that a single request may end up doing all the re-initialization) or when checking a tenant id returned from other static resolvers, or indeed, if the attempt to create a dynamic tenant fails. Dynamic tenants are not even recorded right now when the provided `OidcTenantConfig` fails to resolve, but may be they should be as non-ready as well. which is why I'm proposing to support optional unlimited init retries for all types of tenants (eventually).

IMHO if we start supporting more than 2 initialization retries for static tenants or more than 1 for dynamic ones, we should have a single option for all, and we can wait for this one to be requested.  


> We should retry just once after initialization on startup failed

IMHO we can indeed restrict it to 2 attempts for now for static tenants but as always we can enhance things further :-).